android-components/components/browser/state/src/main/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddleware.kt

```diff
@@ -0,0 +1,85 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.state.engine.middleware
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import mozilla.components.browser.state.action.BrowserAction
+import mozilla.components.browser.state.action.TranslationsAction
+import mozilla.components.browser.state.action.TranslationsAction.TranslateExpectedAction
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.concept.engine.Engine
+import mozilla.components.concept.engine.translate.TranslationError
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * This middleware is for use with managing any states or resources required for translating a
+ * webpage.
+ */
+class TranslationsMiddleware(
+    private val engine: Engine,
+    private val scope: CoroutineScope,
+) : Middleware<BrowserState, BrowserAction> {
+    private val logger = Logger("TranslationsMiddleware")
+
+    override fun invoke(
+        context: MiddlewareContext<BrowserState, BrowserAction>,
+        next: (BrowserAction) -> Unit,
+        action: BrowserAction,
+    ) {
+        // Pre process actions
+        when (action) {
+            is TranslateExpectedAction -> {
```

Comment:
Right now, this action is sent whenever the engine `anticipates` a translation is likely. (It is criteria like the page isn't in the user's preferred language.) This action is the same one for showing the translate icon on the toolbar.

I think this is good to get started with, but later we will need to either listen for an additional action or adjust this one for `unanticipated` translations. Users can also request a translation at any time from the browser menu.

Reply:
Filed [bug 1873672](https://bugzilla.mozilla.org/show_bug.cgi?id=1873672). Totally open to fixing it here too, especially if there is some other pre-existing event that you know off-hand that makes sense.